exchanges/graphcache/src/cacheExchange.ts

```diff
@@ -244,7 +245,59 @@ export const cacheExchange = (opts?: CacheExchangeOpts): Exchange => ({
     let writeDependencies: Set<string> | void;
     let queryDependencies: Set<string> | void;
     if (data !== null && data !== undefined) {
-      writeDependencies = write(store, operation, data).dependencies;
+      /**
+       * Q1 - Q2 (operation series)
+       *
+       * When Q2 arrives first in this dispatch series we can't assume Q1
+       * arriving won't goof up the data in our cache. That's why we'll write
+       * this to an optimistic layer which we'll merge into the main data
+       * after Q1 has arrived.
+       */
+      if (isQuery) {
+        const inFlight: number[] = [...inFlightKeys];
+        // Get the index for the current query, this will be used to opt for
+        // write/layer.
+        let index = inFlight.indexOf(result.operation.key);
+        // When we're at the front of the queue it's safe to write this query to
+        // our permanent data.
+        if (index === 0) {
+          writeDependencies = write(store, operation, data).dependencies;
+          inFlightKeys.delete(operation.key);
+
+          /**
+           * Now we are in a potential secenario where we've build up a series
+           * of results for our operations, for instance a later operation could have
+           * returned earlier than the current one at index 0.
+           *
+           * That's why we'll have to traverse from index 1 up until the first operation
+           * that hs no optimistic-layer attached to it.
+           */
+          for (index = 1; index < inFlight.length; index++) {
+            const opKey = inFlight[index];
+            if (hasOptimisticKey(store.data, opKey)) {
+              mergeOptimisticLayerToData(store.data, opKey);
+              clearOptimistic(store.data, opKey)
+              inFlightKeys.delete(opKey);
+            } else {
+              // Short-circuit this will make it so that we have an uncompleted operation
+              // at index 0 and a set of operations who are potentially completed or uncompleted
+              // behind it.
+              index = inFlight.length;
+            }
+          }
+        } else {
+          /**
+           * When we're not a tthe front of the queue we need to treat this update as an
+           * optimistic layer since this update could always be subject to overriding because
+           * of things earlier in the queue.
+           */
+          writeDependencies = write(store, operation, data, operation.key).dependencies;
```

Comment:
A potential problem here is the following 1 - 2 - 3 where 3 comes in first, applies layer, 2 comes in, applies layer. This will make it so that the data of 2 has prio over 3. This fixes itself when 1 resolves